nvme.c

```diff
@@ -5373,6 +5454,676 @@ static int show_registers(int argc, char **argv, struct command *cmd, struct plu
+	switch (offset) {
+	case NVME_REG_INTMS:
+		if (argconfig_parse_seen(opts, "intms"))
+			flag = true;
+		break;
+	case NVME_REG_INTMC:
+		if (argconfig_parse_seen(opts, "intmc"))
+			flag = true;
+		break;
+	case NVME_REG_CC:
+		if (argconfig_parse_seen(opts, "cc"))
+			flag = true;
+		break;
+	case NVME_REG_CSTS:
+		if (argconfig_parse_seen(opts, "csts"))
+			flag = true;
+		break;
+	case NVME_REG_NSSR:
+		if (argconfig_parse_seen(opts, "nssr"))
+			flag = true;
+		break;
+	case NVME_REG_AQA:
+		if (argconfig_parse_seen(opts, "aqa"))
+			flag = true;
+		break;
+	case NVME_REG_ASQ:
+		if (argconfig_parse_seen(opts, "asq"))
+			flag = true;
+		break;
+	case NVME_REG_ACQ:
+		if (argconfig_parse_seen(opts, "acq"))
+			flag = true;
+		break;
+	case NVME_REG_BPRSEL:
+		if (argconfig_parse_seen(opts, "bprsel"))
+			flag = true;
+		break;
+	case NVME_REG_CMBMSC:
+		if (argconfig_parse_seen(opts, "cmbmsc"))
+			flag = true;
+		break;
+	case NVME_REG_NSSD:
+		if (argconfig_parse_seen(opts, "nssd"))
+			flag = true;
+		break;
+	case NVME_REG_PMRCTL:
+		if (argconfig_parse_seen(opts, "pmrctl"))
+			flag = true;
+		break;
+	case NVME_REG_PMRMSCL:
+		if (argconfig_parse_seen(opts, "pmrmscl"))
+			flag = true;
+		break;
+	case NVME_REG_PMRMSCU:
+		if (argconfig_parse_seen(opts, "pmrmscu"))
+			flag = true;
+		break;
+	default:
+		break;
+	}
+
+	if (flag && cfg->offset == offset) {
+		nvme_show_error("offset duplicated param");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static inline bool nvme_is_write_reg(int offset)
+{
+	switch (offset) {
+	case NVME_REG_INTMS:
+	case NVME_REG_INTMC:
+	case NVME_REG_CC:
+	case NVME_REG_CSTS:
+	case NVME_REG_NSSR:
+	case NVME_REG_AQA:
+	case NVME_REG_ASQ:
+	case NVME_REG_ACQ:
+	case NVME_REG_BPRSEL:
+	case NVME_REG_BPMBL:
+	case NVME_REG_CMBMSC:
+	case NVME_REG_NSSD:
+	case NVME_REG_PMRCTL:
+	case NVME_REG_PMRMSCL:
+	case NVME_REG_PMRMSCU:
+		return true;
+	default:
+		break;
+	}
+
+	return false;
+}
+
+static int set_register_offset_check(struct argconfig_commandline_options *opts,
+				     struct set_reg_config *cfg)
+{
+	int err;
+	int offset;
+
+	if (!argconfig_parse_seen(opts, "value")) {
+		nvme_show_error("value required param");
+		return -EINVAL;
+	}
+
+	for (offset = NVME_REG_CAP; offset <= NVME_REG_PMRMSCU; offset += get_reg_size(offset)) {
+		if (!nvme_is_write_reg(offset))
+			continue;
+		err = set_register_check(opts, cfg, offset);
+		if (err)
+			return err;
+	}
+
+	return 0;
+}
+
+static int set_register_offset(int fd, void *bar, struct argconfig_commandline_options *opts,
+			       struct set_reg_config *cfg)
+{
+	bool set = cfg->offset >= 0;
+	int err;
+	int offset;
+	uint64_t value;
+
+	if (set) {
+		err = set_register_offset_check(opts, cfg);
+		if (err)
+			return err;
+		err = nvme_set_register(fd, bar, cfg->offset, cfg->value, cfg->mmio32);
+		if (err)
+			return err;
+	}
+
+	for (offset = NVME_REG_CAP; offset <= NVME_REG_PMRMSCU; offset += get_reg_size(offset)) {
+		if (!nvme_is_write_reg(offset))
```

Comment:
If we just added  `|| offset == cfg->offset` like `nvme get-reg` does, I think that would avoid the need for the whole `set_register_offset_check()` rigamarole

Reply:
As replied to to the comment https://github.com/linux-nvme/nvme-cli/pull/2188#discussion_r1508336601 as below still `set_register_offset_check()` required for checking if the `--offset` and the `--<register-simbol>` register offsets duplicated.
This check is for example the `--offset` specified as `0x14` CC register offset then `--value` value used to set but if the `--cc` also specified the register set to the option value as below so duplicated the register settings and needed to prevent to set.
```sh
nvme set-reg /dev/nvme0 --offset=0x14 --value=0x460000 --cc=0x350001
```